---
title: "Tutorial"
author: "Yuchen Liao"
date: "6/28/2022"
output: html_document
runtime: shiny
---


## User guide and tutorial

<span style="color:black;">**IMPACT**</span> is a web-based platform for the exploration of <span style="color:black;text-decoration: underline;">**IM**</span>munotherapeutic <span style="color:black;text-decoration: underline;">**P**</span>redictive <span style="color:black;text-decoration: underline;">**A**</span>nd <span style="color:black;text-decoration: underline;">**C**</span>ancer prognos<span style="color:black;text-decoration: underline;">**T**</span>ic biomarkers using genomic, transcriptomic and proteomics datasets from in-house, literature and public databases (TCGA, CPTAC, ICGC, GEO). IMPACT provides versatile functions for analyses about predictive and prognostic biomarkers , including survival analysis, cox regression, subgroup analysis, interaction analysis, tumor mutational landscape analysis, immunogenicity analysis, tumor microenvironment (TME) correlation analysis, and TME comparative analysis. All the generated figures and raw data are freely available to the broad research community (accessibility of in-house data may need further requests). Different functions can be accessed via the menu bar (the green strip as shown below) at the top of the webpage. Detailed function description and tutorial are as follows:

![Home](tutorial-0.png)


<!--
* [User guide and tutorial](#guide) -->
+ [PredExplore: Immunotherapy Predictive Biomarkers Exploration](#predexplore)
+ [ProgExplore: Cancer Prognostic Biomarkers Exploration](#progexplore)
+ [Survival analysis: Kaplan-Meier Curve](#kmcurve)
+ [Survival analysis: Cox regression](#cox)
+ [Survival analysis: Subgroup Analyses](#subgroup)
+ [Interaction Analyses](#interaction)
+ [Immunogenicity Analyses](#immunogenicity)
+ [Tumor microenvironment: TME Correlation Analyses](#tme_corr)
+ [Tumor microenvironment: TME Comparative Analyses](#tme_comp)
+ [MutationProfiles](#mutationprofile)
+ [Data availability](#data)

<!-- 注释
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
-->

--------------------------------------------------------------------------------

<a name="predexplore"> </a>
### PredExplore: Immunotherapy Predictive Biomarkers Exploration

PredExplore allows users to explore the value of interested genes as immunotherapy biomarkers across multiple cohorts. Here, we included 18 cohorts of 11 cancer types. As shown in **Figure 1a** and **1b**, users need to fill in the gene symbols of interest and select a cancer type from the list. The cohort will be stratified into the mutated and wild-type groups based on the queried genes. Then, the results of Kaplan-Meier curves (PFI and/or OS), boxplots (TMB level, Neoantigen expression, and PD-L1 expression), and forest plots of meta analysis for PFS/OS will be partially or fully generated depending on the available information (as shown in **Figure 1c** and **1e**). When one record in the result list (**Figure 1c**) is clicked, corresponding figures will be shown in **Figure 1e**. The generated results and figures can be downloaded by clicking the buttons shown in **Figure 1d**.

<figure style="display:block;margin-left:auto;margin-right:auto;vertical-align:middle;width:60%;">

<img alt="PredExplore" title="PredExplore" src="tour_icip.png" style="display:block;margin-left:auto;margin-right:auto;vertical-align:middle;width:60%;max-width:900px">

<figcaption style="text-align:center;"><b>Figure 1. The interface of Immunotherapy Predictive Biomarkers Exploration</b>. <b>(a)</b> Textbox to fill in multiple interested genes or select from the list. <b>(b)</b> Dropdown list to specify the cancer types. <b>(c )</b> Table of analysis results. <b>(d)</b> Buttons for result downloading. <b>(e)</b> Figures corresponding to the selected item in the result list.</figcaption>

</figure>

--------------------------------------------------------------------------------

<a name="progexplore"> </a>
### ProgExplore: Cancer Prognostic Biomarkers Exploration

By using ProgExplore, users can quickly investigate the prognostic value of interested genes across 40 cohorts of 19 cancer types. Before conducting the investigation, users need to specify a cancer type and fill in the gene symbols of interest in **Figure 2a** and **2b**. Then, the analysis will be launched by clicking the ‘Submit’ button. Detailed results will be shown as a table (**Figure 2c**), and corresponding figures (**Figure 2e**) will be shown by clicking the records in the result list. The generated results and figures can be downloaded by clicking the buttons shown in **Figure 2d**.

<figure style="display:block;margin-left:auto;margin-right:auto;vertical-align:middle;width:60%;">

<img alt="ProgExplore" title="ProgExplore" src="tour_prog.png" style="display:block;margin-left:auto;margin-right:auto;vertical-align:middle;width:60%;max-width:900px">

<figcaption style="text-align:center;"><b>Figure 2. The interface of Cancer Prognostic Biomarker Exploration</b>. <b>(a)</b> Textbox to fill in multiple interested genes or select from the list. <b>(b)</b> The drop-down list of specifying the cancer type. <b>(c )</b> Table of detailed analysis results. <b>(d)</b> Buttons for result downloading. <b>(e)</b> Figures corresponding to the selected record in the result list.</figcaption>

</figure>

--------------------------------------------------------------------------------

<a name="kmcurve"> </a>
### Survival analysis: Kaplan-Meier Curve

Kaplan-Meier curve is used to estimate the prognostic value of mutation status or expression alteration of interested genes. As shown in **Figure 3a**, users need to select the study type, cancer type and research cohort, fill in the genes of interest and choose the mutation type or select the cutoff values for expression data to launch the analysis by clicking the ‘Submit’ button. Users are also allowed to customize the colors in the figure. For the results, users can select the Kaplan-Meier curve of different genes by clicking the tab on top of **Figure 3b**. The ‘All Queried Genes’ tab shows the combined overall alteration results of the selected gene set (the combined result will mark the sample as mutated if more than one gene of the gene set is mutated). The ‘All Queried Genes’ tabs appeared in the following context or figures share the same meaning here. Detailed results for Kaplan-Meier curve plotting are listed in **Figure 3d**. Moreover, all the corresponding results and figures can be downloaded by clicking the buttons shown in **Figure 3c**.

<figure style="display:block;margin-left:auto;margin-right:auto;vertical-align:middle;width:60%;">

<img alt="Kaplan-Meier Curve" title="Kaplan-Meier Curve" src="tour_km.png" style="display:block;margin-left:auto;margin-right:auto;vertical-align:middle;width:60%;max-width:900px">

<figcaption style="text-align:center;"><b>Figure 3. The interface of survival analysis with Kaplan-Meier estimation</b>. <b>(a)</b> Control panel to select the cancer type, research cohort, and mutation types, and fill in multiple interested genes. <b>(b)</b> Kaplan-Meier curves of selected genes and the combined results. <b>(c )</b> Buttons for result downloading. <b>(d)</b> Table of detailed data for Kaplan-Meier curve plotting.</figcaption>

</figure>

--------------------------------------------------------------------------------

<a name="cox"> </a>
### Survival analysis: Cox regression

Cox regression is used to evaluate the prognostic value of the gene alteractions/expression levels and clinical properties, including age, sex, stage, smoking, and TMB et al. To start the analysis, users need to input the genes of interest and select the study type, a cancer type, a research cohort, clinical properties, and mutation types (as shown in **Figure 4a**) or cutoff values. Univariable and multivariable Cox regression will be conducted using clinical properties with the mutation or expression information of selected genes. Forest plots and detailed results will be generated, as shown in **Figure 4b** and **4d**, respectively. All the results and figures can also be downloaded by clicking the download buttons in **Figure 4c**.

<figure style="display:block;margin-left:auto;margin-right:auto;vertical-align:middle;width:60%;">

<img alt="Cox regression" title="Cox regression" src="tour_cox.png" style="display:block;margin-left:auto;margin-right:auto;vertical-align:middle;width:60%;max-width:900px">

<figcaption style="text-align:center;"><b>Figure 4. The interface of survival analysis using cox regression</b>. <b>(a)</b> Control panel to select the study type, cancer type, research cohort, and mutation types, and fill in multiple interested genes. <b>(b)</b> Forest plot of cox regression using selected genes and clinical properties. <b>(c )</b> Buttons for result downloading. <b>(d)</b> Table of detailed results of Cox regression.</figcaption>

</figure>

--------------------------------------------------------------------------------

<a name="subgroup"> </a>
### Survival analysis: Subgroup Analyses

Subgroup analysis aims to investigate the interaction effects of mutation status or expression of genes of interest with related clinical characteristics. As shown in **Figure 5a**, users need to input the genes of interest and select related options of the study type, a study cohort, interaction factor type, related clinical characteristics, and mutation types (input cutoff values for expression data). **Figures 5b** and **5d** show the results of Subgroup Analysis. Users can check the results of different genes by clicking the tabs of gene symbols. Detailed result table can be downloaded via the button in **Figure 5c**.


<figure style="display:block;margin-left:auto;margin-right:auto;vertical-align:middle;width:60%;">

<img alt="Subgroup Analyses" title="Subgroup Analyses" src="tour_subgroup.png" style="display:block;margin-left:auto;margin-right:auto;vertical-align:middle;width:60%;max-width:900px">

<figcaption style="text-align:center;"><b>Figure 5. The interface for Subgroup analyses</b>. <b>(a)</b> Control panel to select the study type, cancer type, research cohort, the interaction factor type, clinical characteristics, and mutation types, and input multiple interested genes. <b>(b)</b> Forest plot and detailed analysis results. <b>(c )</b> Buttons for result downloading. <b>(d)</b> Table of detailed results.</figcaption>

</figure>

--------------------------------------------------------------------------------

<a name="interaction"> </a>
### Interaction Analyses

In the Interaction Analysis, we provide the function of testing for interaction effects of gene mutation status (or expression levels) with treatment. To start the analysis, users need to fill in the gene symbols of interest in the textbox and select the mutation types (input the cutoff values for expression data) in the control panel (**Figure 6a**). **Figures 6b** and **6d** show the analysis results of Kaplan-Meier curves and detailed result table, respectively. The value of P for Interaction is shown in Figure 6b. Users can also download the figures and results via the buttons in **Figure 6c**.

<figure style="display:block;margin-left:auto;margin-right:auto;vertical-align:middle;width:60%;">

<img alt="Interaction Analyses" title="Interaction Analyses" src="tour_interaction.png" style="display:block;margin-left:auto;margin-right:auto;vertical-align:middle;width:60%;max-width:900px">

<figcaption style="text-align:center;"><b>Figure 6. The interface for Interaction analysis</b>. <b>(a)</b> Control panel to fill in multiple interested genes and select mutation types and visualization colors. <b>(b)</b> Kaplan-Meier curves of the analysis results. <b>(c )</b> Buttons for result downloading. <b>(d)</b> Table of detailed results.</figcaption>

</figure>

--------------------------------------------------------------------------------

<a name="immunogenicity"> </a>
### Immunogenicity Analyses

In Immunogenicity analysis, we collected the continuous immune-related biomarkers and performed the comparative analysis of tumor purity, TMB, NeoAntigen, and PD-L1 expression based on the mutation status or expression levels of selected biomarkers. To start the analysis, users need to fill in the genes of interest and select the study type, a cancer type, a research cohort, data type, clinical properties, and mutation types (as shown in **Figure 7a**) or cutoff values for expression data. Then, the comparative analysis will be conducted using the mutation information of selected genes. Boxplots and detailed results will be generated as shown in Figures **7b** and **7d**, respectively. The results and figures can also be downloaded by clicking the download buttons in Figure **7c**.

<figure style="display:block;margin-left:auto;margin-right:auto;vertical-align:middle;width:60%;">

<img alt="Immunogenicity Analyses" title="Immunogenicity Analyses" src="tour_genicity.png" style="display:block;margin-left:auto;margin-right:auto;vertical-align:middle;width:60%;max-width:900px">

<figcaption style="text-align:center;"><b>Figure 7. The interface of immunogenicity analysis</b>. <b>(a)</b> Control panel to select the study type, cancer type, research cohort, and mutation types, and fill in multiple interested genes. <b>(b)</b> Boxplot of TMB, Neoantigen, and PDL1-exp difference of the mutated and wild-type groups. <b>(c )</b> Buttons for result downloading. <b>(d)</b> Table of detailed results.</figcaption>

</figure>

--------------------------------------------------------------------------------

<a name="tme_corr"> </a>
### Tumor microenvironment: TME Correlation Analyses

In the TME Correlation Analysis, users can investigate the correlations of selected genes with selected immunomodulator or signature, including gene sets (Chemokine, Receptor, MHC, Immunoinhibitor, Immunostimulator et al.) and pathways (Immune-related signatures, Key signatures, Oncogenic Signalling et al.). To conduct the analysis, users need to select the study type, a cancer type, a research cohort, an immunomodulator type, a correlation algorithm, and the visualization type. Then, genes of interest are needed to fill in the textbox in **Figure 8a**. Colors of different groups can also be customized by changing the values of ‘Color 1’ and ‘Color 2’ in **Figure 8a**. Correlation analysis results can be found in **Figures 8b** and **8d**. The results and figures can be downloaded by clicking the buttons in **Figure 8c**.

<figure style="display:block;margin-left:auto;margin-right:auto;vertical-align:middle;width:60%;">

<img alt="TME Correlation Analyses" title="TME Correlation Analyses" src="tour_corr.png" style="display:block;margin-left:auto;margin-right:auto;vertical-align:middle;width:60%;max-width:900px">

<figcaption style="text-align:center;"><b>Figure 8. The interface of TME Correlation analysis</b>. <b>(a)</b> Control panel to select the study type, cancer type, research cohort, correlation algorithm, visualization type, and fill in multiple interested genes. <b>(b)</b> Bubble plot or Heatmap of the correlation analysis result. <b>(c )</b> Buttons for result downloading. <b>(d)</b> Table of detailed results.</figcaption>

</figure>

--------------------------------------------------------------------------------

<a name="tme_comp"> </a>
### Tumor microenvironment: TME Comparative Analyses

In the TME Comparative Analysis function, users can easily investigate the immunomodulator or signature differences in low- and high-expressed groups of selected genes. Users need to select the study type, a cancer type, a research cohort, and an immunomodulator/signature type. After that, genes of interest and cutoff values need to be given in corresponding textboxes (**Figure 9a**). Colors of different groups can also be customized by changing the values of ‘Color 1’ and ‘Color 2’ in **Figure 9a**. Heatmap and detailed results of Comparative analysis can be found in **Figures 9b** and **9d**. Users are allowed to download the results and figures by clicking the buttons in **Figure 9c**.

<figure style="display:block;margin-left:auto;margin-right:auto;vertical-align:middle;width:60%;">

<img alt="TME Comparative Analyses" title="TME Comparative Analyses" src="tour_comp.png" style="display:block;margin-left:auto;margin-right:auto;vertical-align:middle;width:60%;max-width:900px">

<figcaption style="text-align:center;"><b>Figure 9. The interface of the TME Comparative Analysis</b>. <b>(a)</b> Control panel to select the study type, cancer type, research cohort, immonomodulator/signature type, and fill in multiple interested genes and cutoff values. <b>(b)</b> Heatmap of the TME Comparative Analysis result. <b>(c )</b> Buttons for result downloading. <b>(d)</b> Table of detailed results.</figcaption>

</figure>

--------------------------------------------------------------------------------

<a name="mutationprofile"> </a>
### MutationProfiles

Tumor Mutation Profile function helps to visualize the mutation status of selected genes in groups stratified by the mutation status or expression levels of interested genes. To visualize the mutation status, users need to select the study type, a cancer type, and mutation types (**Figure 10a**) or cutoff values for expression data. Then, genes for stratification and genes for visualization are needed to fill in the corresponding textboxes separately (**Figure 10a**). Colors of different groups can also be customized by changing the values of ‘Color 1’ and ‘Color 2’ in **Figure 10a**. Results are shown in **Figures 10b** and **10d**. The results and figures can be downloaded by clicking the buttons in **Figure 10c**.

<figure style="display:block;margin-left:auto;margin-right:auto;vertical-align:middle;width:60%;">

<img alt="MutationProfiles" title="MutationProfiles" src="tour_mutation.png" style="display:block;margin-left:auto;margin-right:auto;vertical-align:middle;width:60%;max-width:900px">

<figcaption style="text-align:center;"><b>Figure 10. The interface of Tumor mutation profile visualization</b>. <b>(a)</b> Control panel to select the study type, cancer type, research cohort, and mutation types, and fill in multiple interested genes. <b>(b)</b> Fall-plot of the mutation status of selected genes in different groups. <b>(c )</b> Buttons for result downloading. <b>(d)</b> Table of detailed results.</figcaption>

</figure>

--------------------------------------------------------------------------------

<a name="data"> </a>
### Data availability

All the datasets used in IMPACT are listed as a table on the ‘Data’ page. Detailed information on each dataset can be found in the table as shown in **Figure 11**. We provide the download links for all the publicly available datasets. Further requests may be needed to download our in-house datasets..


<figure style="display:block;margin-left:auto;margin-right:auto;vertical-align:middle;width:60%;">

<img alt="Data availability" title="Data availability" src="tour_data.png" style="display:block;margin-left:auto;margin-right:auto;vertical-align:middle;width:60%;max-width:900px">

<figcaption style="text-align:center;"><b>Figure 11. The screenshot of dataset information used in IMPACT</b>.</figcaption>

</figure>

